# (실습) 넘파이 어레이

**주의사항**

- `None`은 적절한 표현식으로 대체한다.
- `Pass`는 적절한 명령문으로 대체한다.

## numpy 모듈

**문제 1**

아래 코드를 설명하라.

In [ ]:
import numpy as np

`코드설명`:

아래 코드는 수정하지 마세요!!!

In [ ]:
np.__version__

'1.26.4'

**문제 2**

아래 코드의 기능을 설명하라

In [ ]:
np.random.seed(42)

`코드설명`: 난수를 생성한다.

**문제 3**

부동소수점으로 구성된 어레이의 항목은 기본적으로 소수점 이하 8자리까지 출력한다.

In [ ]:
Z = np.random.random((40,40))
Z

array([[0.37454012, 0.95071431, 0.73199394, ..., 0.09767211, 0.68423303,
        0.44015249],
       [0.12203823, 0.49517691, 0.03438852, ..., 0.07404465, 0.35846573,
        0.11586906],
       [0.86310343, 0.62329813, 0.33089802, ..., 0.18657006, 0.892559  ,
        0.53934224],
       ...,
       [0.81783422, 0.12020906, 0.5444891 , ..., 0.89803055, 0.2434823 ,
        0.92703454],
       [0.06026739, 0.93443603, 0.35162269, ..., 0.41102847, 0.8398613 ,
        0.90002312],
       [0.35342138, 0.23687056, 0.78052552, ..., 0.07941937, 0.08565825,
        0.89419087]])

`np.set_printoptions()` 의 기능을 다음 두 개의 예를 이용하여 설명하라.

In [ ]:
np.set_printoptions(precision=4)

In [ ]:
Z

array([[0.3745, 0.9507, 0.732 , ..., 0.0977, 0.6842, 0.4402],
       [0.122 , 0.4952, 0.0344, ..., 0.074 , 0.3585, 0.1159],
       [0.8631, 0.6233, 0.3309, ..., 0.1866, 0.8926, 0.5393],
       ...,
       [0.8178, 0.1202, 0.5445, ..., 0.898 , 0.2435, 0.927 ],
       [0.0603, 0.9344, 0.3516, ..., 0.411 , 0.8399, 0.9   ],
       [0.3534, 0.2369, 0.7805, ..., 0.0794, 0.0857, 0.8942]])

In [ ]:
np.set_printoptions(precision=6)

In [ ]:
Z

array([[0.37454 , 0.950714, 0.731994, ..., 0.097672, 0.684233, 0.440152],
       [0.122038, 0.495177, 0.034389, ..., 0.074045, 0.358466, 0.115869],
       [0.863103, 0.623298, 0.330898, ..., 0.18657 , 0.892559, 0.539342],
       ...,
       [0.817834, 0.120209, 0.544489, ..., 0.898031, 0.243482, 0.927035],
       [0.060267, 0.934436, 0.351623, ..., 0.411028, 0.839861, 0.900023],
       [0.353421, 0.236871, 0.780526, ..., 0.079419, 0.085658, 0.894191]])

`코드설명`: 값이 출력될 때 소수점이 길경우에 np.set_printoptions를 이용하면 출력옵션을 변경할수있고 이때 precision option의 값을 조정하면 소숫점 아래 숫자의 개수(정밀도)를 조정할 수 있다.

아래 코드는 기본 옵션값으로 되돌린다. 수정하지 마세요.

In [ ]:
np.set_printoptions(precision=8)

## 1차원 어레이

**문제 1**

크기가 10이고 모든 항목이 0인 1차원 어레이를
변수 `Z`가 가리키도록 하라.
단, `np.zeros()` 함수를 반드시 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.zeros(shape=(10,), dtype=np.int8)
Z

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int8)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])).all()

**문제 2**

변수 `Z`가 10에서 49까지의 정수 중에서 짝수로만 구성된 1차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수를 반드시 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.arange(10, 49, 2)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42,
                        44, 46, 48])).all()

**문제 3**

변수 `Z`가 아래와 같이 선언되었다.

In [ ]:
Z = np.array([2, 3, 1, 0, 4])
Z

array([2, 3, 1, 0, 4])

변수 `Z`의 항목을 정렬하라.
단, sort() 메서드를 활용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z.sort()
Z

array([0, 1, 2, 3, 4])

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.arange(5)).all()

**문제 4**

변수 `Z`가 크기 30이며 무작위로 생성된 1차원 어레이를 가리키도록 하라.
또한 포함된 항목들의 평균값을 계산하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
`np.random.random()` 함수, `mean()` 어레이 메서드를 반드시 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random(30)
Z_mean = np.mean(Z)
Z

array([0.38062329, 0.87797432, 0.86805669, 0.8059254 , 0.79003044,
       0.30467914, 0.08091928, 0.40298018, 0.17352451, 0.69495109,
       0.34609973, 0.9756102 , 0.64097208, 0.82248056, 0.13252467,
       0.86201448, 0.92275719, 0.48706192, 0.60625294, 0.7648098 ,
       0.17483863, 0.50256608, 0.39866303, 0.14637399, 0.36753442,
       0.06817231, 0.02581191, 0.13516629, 0.96311511, 0.54952954])

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z.shape == (30,))
print(f"1차원 어레이 평균값: {Z_mean}")

1차원 어레이 평균값: 0.5090673071853106


**문제 5**

1차원 어레이가 다음과 같이 주어졌다.

In [ ]:
Z = np.arange(1, 4)
Z

array([1, 2, 3])

(1) 아래 표현식의 계산 결과를 설명하라.

In [ ]:
Z*Z

array([1, 4, 9])

`코드설명`: arange(1, 4)는 1에서 4까지에서 1을 시작으로 1간격씩 떨어진 수를 array에 집어넣는다 이때 4는 포함하지 않기에 array([1, 2, 3])이 되고 두번 곱하게 되면 array([1, 4, 9])가 된다.

(2) 아래 표현식의 계산 결과를 설명하라.

In [ ]:
Z**Z

array([ 1,  4, 27])

`코드설명`: Z**Z의 경우 지수승에 대한 거듭제곱이므로 array([1, 2, 3])에 대하여 1에 1제곱/ 2에 2제곱/ 3에 3제곱이 되어 array([1, 4, 27])로 출력된다.

(3) 아래 표현식의 계산 결과를 설명하라.

In [ ]:
Z < -Z

array([False, False, False])

`코드설명`: array의 값에 -를 붇혔을때 값이 크다라는 조건에 대하여 맞을 경우엔 True를 아닐 경우엔 False를 반환한다.

(4) 아래 표현식의 계산 결과를 설명하라.

In [ ]:
Z/1/1

array([1., 2., 3.])

`코드설명`: array([1, 2, 3])에 1로 두번 나누기 때문에 값은 변하는거 없이 array([1, 2, 3])이다.

**문제 6**

변수 `Z`가 아래 모양의 어레이를 가리키도록 하라.
단, `np.linspace()` 함수와 `endpoint` 키워드 인자를 이용한다.

```
[0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
```

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.linspace(0, 1, 10, endpoint=False)
Z

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])

아래 코드는 수정하지 마세요!!!

In [ ]:
assert np.isclose(Z, np.array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])).all()

**문제 7**

아래 코드의 실행결과를 설명하라.

In [ ]:
Z = np.zeros((10,10))
#print(Z.size)
#print(Z.itemsize)
print("%d bytes" % (Z.size * Z.itemsize))

800 bytes


`코드설명`: np.zero()를 이용하여 10 * 10 크기를 0으로 채운 배열을 생성한다 이때 배열 Z의 전체 사이즈는 10 * 10 = 100이고 각각의 요소의 사이즈는 int8 형이므로 8이다.

**문제 8**

다음 리스트에서 0이 아닌 항목들의 인덱스로 구성된 어레이를 한줄 코드로 생성하라.
단. `np.nonzero()` 함수를 활용한다.

In [ ]:
Z = [1, 2, 0, 0, 4, 0]

In [ ]:
# None을 적절한 표현식으로 대체하라.

non_zeros = np.nonzero(Z)
non_zeros

(array([0, 1, 4]),)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (non_zeros == np.array([0, 1, 4])).all()

**문제 9**

두 개의 1차원 어레이가 다음과 같이 주어졌다.

In [ ]:
A = np.ones(3)*1
B = np.ones(3)*2

아래 코드의 실행 결과를 설명하라.

In [ ]:
(A+B)*(-A/2)

array([-1.5, -1.5, -1.5])

`코드설명`: A 배열의 경우 3개의 요소가 전부 1로 되있는 배열을 생성 후 전부에 1을 곱해주고 B 배열의 경우 3개의 요소가 전부 1로 되있는 배열을 생성 후 전부에 2를 곱해준다. 그 뒤에 (A+B)*(-A/2)를 계산하게 되면 ([3, 3, 3]) * ([-0.5, -0.5, -0.5]) = ([-1.5, -1.5, -1.5])가 된다.

In [ ]:
np.add(A,B,out=B)

array([3., 3., 3.])

`코드설명`: 배열 A = ([1, 1, 1])와 배열 B = ([2, 2, 2])를 더한 값 ([3, 3, 3])을 배열 B에 값을 넣는다라는 개념이다.

In [ ]:
np.divide(A,2,out=A)

array([0.125, 0.125, 0.125])

`코드설명`: 배열 A=([1,1,1])를 2로 나누기한 ([0.5, 0.5, 0.5])를 배열 A에 값을 넣는다라는 개념이다.

In [ ]:
np.negative(A,out=A)

array([-1., -1., -1.])

`코드설명`: 배열 A=([1, 1, 1])를 부호 반전을 시켜 ([-1, -1, -1])로 만들고 배열 A에 값을 넣는다라는 개념이다.

In [ ]:
np.multiply(A,B,out=A)

array([2., 2., 2.])

`코드설명`: 배열 A = ([1, 1, 1])와 배열 B = ([2, 2, 2])를 곱한 값 ([2, 2, 2])를 배열 A에 값을 넣는다라는 개념이다.

## 2차원 어레이

**문제 1**

변수 `Z`가 (5, 5) 모양의 2차원 어레이를 가리키도록 하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
또한 `np.random.random()` 함수를 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random((5, 5))
Z

array([[0.26113976, 0.15300002, 0.4112904 , 0.02141841, 0.72211365],
       [0.40891301, 0.15024048, 0.87613241, 0.13996508, 0.04498156],
       [0.75683758, 0.57156924, 0.02166819, 0.92995183, 0.12653454],
       [0.36933661, 0.99845646, 0.45628392, 0.26542162, 0.25128003],
       [0.9981387 , 0.7089722 , 0.41102133, 0.44976923, 0.51959067]])

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z.shape == (5, 5))
assert (Z.dtype == np.float64)

**문제 2**

변수 `Z`가 0에서 8까지의 정수로 이루어진 (3, 3) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수와 `reshape()` 어레이 메서드를 반드시 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = None

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.array([[0, 1, 2],
                       [3, 4, 5],
                       [6, 7, 8]])).all()

**문제 3**

변수 `Z`가 (3, 3) 모양의 2차원 어레이를 가리키도록 하라.
단, 2차원 어레이의 대각선 상은 1이고 나머지는 0이어야 한다.
또한 `np.eye()` 함수를 반드시 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = None

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.array([[1., 0., 0.],
                       [0., 1., 0.],
                       [0., 0., 1.]])).all

**문제 4**

변수 `Z`가 아래와 같은 (5,5) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수와 `np.diag()` 함수의 `k` 키워드 인자를 반드시 활용한다.

```
[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]
```

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = None

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.array([[0, 0, 0, 0, 0],
                        [1, 0, 0, 0, 0],
                        [0, 2, 0, 0, 0],
                        [0, 0, 3, 0, 0],
                        [0, 0, 0, 4, 0]])).all()

**문제 5**

변수 `Z`가 아래와 같은 (5, 5) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.zeros()` 함수, `np.arange()` 함수, 2차원 어레이 덧셈을 이용한다.

```
[[0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]]
```

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.zeros((5,5)) + np.arange(5)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.array([[0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.]])).all()

**문제 6**

아래 3x3 단위 행렬에 해당하는 2차원 어레이를 한줄 코드로 생성하라.
단, `np.eye()` 함수를 이용한다.

```
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
```

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.eye(3)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z == np.array([[1., 0., 0.],
                       [0., 1., 0.],
                       [0., 0., 1.]])).all()

**문제 7**

무작위 수로 구성된 (5, 5) 모양의 2차원 어레이가 다음과 같다.

In [ ]:
# 지정한 범위로 값을 조정하는 것 : 정규화, -1 ~ 1, 0.5
# 대원칙 : 인공지능은 학습전에 데이터가 정규화/표준화가 되어있어야 한다.
Z = np.random.random((5, 5))

변수 `Z`가 가리키는 2차원 어레이에 대해 min-max 크기조정(정규화)을 적용하라.

min-max 크기조정(정규화)은 항목의 모든 값을 아래 식을 통해 0과 1 사이의 부동소수점으로 변환한다. $max$와 $min$은 각각 최대 항목과 최소 항목을 가리킨다.

$$
\frac{x-min}{max - min}
$$

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = (Z - Z.min()) / (Z.max() - Z.min())
Z

array([[1.        , 0.79298109, 0.53737814, 0.46640693, 0.13864297],
       [0.49230565, 0.14471449, 0.4149446 , 0.        , 0.06456492],
       [0.40506832, 0.09221994, 0.92979822, 0.03999708, 0.11263059],
       [0.43976923, 0.95344452, 0.00166555, 0.94738159, 0.46444661],
       [0.93474605, 0.44005011, 0.74750225, 0.14168011, 0.5746975 ]])

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z.max() == 1 and Z.min() == 0)

## 3차원 어레이

**문제 1**

(1) 변수 `Z`가 (3, 3, 3) 모양의 3차원 어레이를 가리키도록 하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
또한 `np.random.random()` 함수를 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random((3,3,3))
Z

array([[[0.21679276, 0.1505768 , 0.57895278],
        [0.58067912, 0.88527347, 0.26794932],
        [0.15537237, 0.98985108, 0.25069061]],

       [[0.5222168 , 0.00668475, 0.82952479],
        [0.74865564, 0.57304496, 0.4512036 ],
        [0.13652892, 0.33593444, 0.72502894]],

       [[0.13494221, 0.31670825, 0.71179237],
        [0.05046885, 0.0058605 , 0.40428861],
        [0.31301918, 0.65474503, 0.81863528]]])

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z.shape == (3, 3, 3))
assert (Z.dtype == 'f8')

(2) `Z`에 포함된 항목들 중에서 최댓값과 최솟값의 차이를 구하라.
단, `maximum`, `mininum` 두 변수가 각각 최댓값과 최솟값을 가리키도록 하며,
`max()`와 `min()` 어레이 메서드를 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
maximum = Z.max()
minimum = Z.min()

아래 코드는 수정하지 마세요!!!

In [ ]:
print(f"최댓값과 최솟값의 차이: {maximum - minimum}")

최댓값과 최솟값의 차이: 0.9839905813632115


## `dtype` 종류

**문제 1**

아래 두 코드의 실행결과를 설명하라.

In [ ]:
for dtype in [np.int8, np.int32, np.int64]:
    print(np.iinfo(dtype).min)
    print(np.iinfo(dtype).max)

-128
127
-2147483648
2147483647
-9223372036854775808
9223372036854775807


`코드설명`: 정수형 자료형에 허용되는 비트수에 따른 최소값과 최대값
8비트 정수형(np.int8) : -128 ~ 127
16비트 정수형(np.int16) : -2147483648 ~ 2147483647
32비트 정수형(np.int32) : -9223372036854775808 ~ 9223372036854775807

In [ ]:
for dtype in [np.float32, np.float64]:
    print(np.finfo(dtype).min)
    print(np.finfo(dtype).max)
    print(np.finfo(dtype).eps)

`코드설명`:

**문제 2**

부동소수점으로 구성된 1차원 어레이가 아래와 같이 주어졌다.

In [ ]:
Z = (np.random.rand(10)*100).astype(np.float32)

변수 `Y`가 변수 `Z`가 가리키는 1차원 어레이의 dtype이 `int32`로 형변환된 1차원 어레이를 가리키도록 하라.
단, `astype()` 메서드를 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Y = None

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Y.shape == (10,))
assert (Y.dtype == np.int32)

**문제 3**

변수 `Z`가 아래와 같이 무작위로 생성된 길이가 10인 1차원 어레이를 가리킨다.

In [ ]:
Z = np.random.uniform(0,10,10)
Z

array([1.86460516, 7.91112971, 4.32932559, 9.98049   , 7.64545077,
       6.76597363, 3.60680678, 8.0246804 , 4.97424797, 7.95016248])

(1) 나머지 연산자(`%`)를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Y = Z -Z%1
Y

array([1., 7., 4., 9., 7., 6., 3., 8., 4., 7.])

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Y == Z.astype(np.int32)).all()

(2) 몫 연산자(`//`)를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Y = Z // 1

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Y == Z.astype(np.int32)).all()

(3) 내림 함수 `np.floor()` 함수를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Y = np.floor(Z)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Y == Z.astype(np.int32)).all()

(4) `np.trunc()` 함수를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Y = np.trunc(Z)

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Y == Z.astype(np.int32)).all()